### DISCUSSION TITLE
[Feature request] Feature-protected whitelist in Eth2NearClient

### DISCUSSION THREAD

**Discussion by maxzaver** [aurora-is-near/rainbow-bridge#477] [Ideas]
# Motivation

We would like to a launch a version of the bridge that allows only certain users to use it. This can be useful for running the bridge on the Mainnet before we make it completely publicly available.

# Design

We would like to modify [block_hash_safe](https://github.com/near/rainbow-bridge/blob/master/near-contracts/eth-client/src/lib.rs#L188) to assert that `env::signer_account_id() ` is in a given list of whitelisted accounts.

We would also like to add methods `add_whitelist_user(user: AccountId)` and `remove_white_list_user(user: AccountId)` which would add or remove users from the whitelist.

We would also hide this code behind a non-default feature flag `whitelist_users`.

Assigning to @Kouprin since he is a TL who can decide who is working on it.

**Comment by mfornet** [comment]
> This can be useful for running the bridge on the Mainnet before we make it completely publicly available.

@nearmax I don't understand the motivation for this request. Is this to protect users from using unstable version of the bridge and loosing money? 

> We would also hide this code behind a non-default feature flag whitelist_users.

If we are having two different binaries, it means we need to deploy two versions of the bridge, which I think defeats the purpose of the feature at all.

My opinion is that once the bridge is deployed in mainnet (which is _almost_ the case right now) it should be available to everyone. The point is that we don't publicly announce it until we consider it is safe to use. Partners that need the bridge asap should be able to use it without the whitelist and being aware of the **DISCLAIMER**.

@djsatok what is your opinion on this?
